icu4c/source/common/unicode/utfiter.h

```diff
@@ -0,0 +1,2045 @@
+public:
+    // Handle ill-formed UTF-8
+    static inline CP32 sub() {
+        switch (behavior) {
+            case UTF_BEHAVIOR_NEGATIVE: return U_SENTINEL;
+            case UTF_BEHAVIOR_FFFD: return 0xfffd;
+        }
+    }
+
+    static inline void inc(UnitIter &p, const UnitIter &limit) {
+        // Very similar to U8_FWD_1().
+        uint8_t b = *p;
+        ++p;
+        if (U8_IS_LEAD(b) && p != limit) {
+            uint8_t t1 = *p;
+            if ((0xe0 <= b && b < 0xf0)) {
+                if (U8_IS_VALID_LEAD3_AND_T1(b, t1) &&
+                        ++p != limit && U8_IS_TRAIL(*p)) {
+                    ++p;
+                }
+            } else if (b < 0xe0) {
+                if (U8_IS_TRAIL(t1)) {
+                    ++p;
+                }
+            } else /* b >= 0xf0 */ {
+                if (U8_IS_VALID_LEAD4_AND_T1(b, t1) &&
+                        ++p != limit && U8_IS_TRAIL(*p) &&
+                        ++p != limit && U8_IS_TRAIL(*p)) {
+                    ++p;
+                }
+            }
+        }
+    }
+
+    static inline void dec(UnitIter start, UnitIter &p) {
+        // Very similar to U8_BACK_1().
+        uint8_t c = *--p;
+        if (U8_IS_TRAIL(c) && p != start) {
+            UnitIter p1 = p;
+            uint8_t b1 = *--p1;
+            if (U8_IS_LEAD(b1)) {
+                if (b1 < 0xe0 ||
+                        (b1 < 0xf0 ?
+                            U8_IS_VALID_LEAD3_AND_T1(b1, c) :
+                            U8_IS_VALID_LEAD4_AND_T1(b1, c))) {
+                    p = p1;
+                    return;
+                }
+            } else if (U8_IS_TRAIL(b1) && p1 != start) {
+                uint8_t b2 = *--p1;
+                if (0xe0 <= b2 && b2 <= 0xf4) {
+                    if (b2 < 0xf0 ?
+                            U8_IS_VALID_LEAD3_AND_T1(b2, b1) :
+                            U8_IS_VALID_LEAD4_AND_T1(b2, b1)) {
+                        p = p1;
+                        return;
+                    }
+                } else if (U8_IS_TRAIL(b2) && p1 != start) {
+                    uint8_t b3 = *--p1;
+                    if (0xf0 <= b3 && b3 <= 0xf4 && U8_IS_VALID_LEAD4_AND_T1(b3, b2)) {
+                        p = p1;
+                        return;
+                    }
+                }
+            }
+        }
+    }
+
+    static inline CodeUnits<CP32, UnitIter> readAndInc(UnitIter &p, UnitIter limit) {
+        // Very similar to U8_NEXT_OR_FFFD().
+        UnitIter p0 = p;
+        CP32 c = uint8_t(*p);
+        ++p;
+        if (U8_IS_SINGLE(c)) {
+            return {c, 1, true, p0, p};
+        }
+        uint8_t length = 1;
+        uint8_t t = 0;
+        if (p != limit &&
+                // fetch/validate/assemble all but last trail byte
+                (c >= 0xe0 ?
+                    (c < 0xf0 ?  // U+0800..U+FFFF except surrogates
+                        U8_LEAD3_T1_BITS[c &= 0xf] & (1 << ((t = *p) >> 5)) &&
```

Comment:
So this `c &= 0xf` here modifies the value of `c` halfway through the `if`-expression!? Just checking that this really is intentional …

Reply:
It is. As noted in comments, these implementations are nearly the same as the C macros, and I took a lot of time optimizing those some years ago.

Compare with https://github.com/unicode-org/icu/blob/release-77-1/icu4c/source/common/unicode/utf8.h#L389